tests/storages_tests/test_storages.py

```diff
@@ -1232,3 +1227,28 @@ def failed_trial_callback(study: "optuna.Study", trial: FrozenTrial) -> None:
         optuna.storages.fail_stale_trials(study)
 
         assert study.trials[0].state is TrialState.FAIL
+
+
+@pytest.mark.parametrize("storage_mode", ["sqlite", "redis"])
+def test_fail_stale_trials_raw(storage_mode: str) -> None:
+    heartbeat_interval = 1
+    grace_period = 2
+
+    with StorageSupplier(storage_mode) as storage:
+        assert isinstance(storage, (RDBStorage, RedisStorage))
+        storage.heartbeat_interval = heartbeat_interval
+        storage.grace_period = grace_period
+
+        study_id = storage.create_new_study()
+        assert storage.fail_stale_trials(study_id) == []
+
+
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES_HEARTBEAT)
```

Comment:
Why the possible `storage_mode`s are `STORAGE_MODES_HEARTBEAT`? Might be `STORAGE_MODES`.

Reply:
```STORAGE_MODES``` は ```inmemory``` を含みますが、```inmemory```はもちろんremote storageではないので```read_trials_from_remote_storage()```を持たず、エラーとなります。

そのため```STORAGE_MODES_HEARTBEAT```を使用しました。「hertbeat と無関係なテストでなぜ```STORAGE_MODES_HEARTBEAT```を使うのだ？」という違和感があるのは理解できます。私もそうだったので、当初は ```"sqlite", "cached_sqlite", "redis", "cached_redis"``` と並べて書こうかなとも思いましたが、結局「定数使え」と言われそうだなと思い、今の形にしました。

Reply:
`STORAGE_MODES` contains `inmemory` storage, but `inmemory` storage is of course not remote storage, so it does not have `read_trials_from_remote_storage()`, which will result in an error.

This is why we used `STORAGE_MODES_HEARTBEAT`. You may be wondering, "Why use `STORAGE_MODES_HEARTBEAT` for a test that has nothing to do with hertbeat?" I can understand why you might feel uncomfortable. At first, I thought about writing `"sqlite", "cached_sqlite", "redis", and "cached_redis"`, but then I thought that people would say "use constants", so I decided to use the current form.

Reply:
Thank yoo for the explanation. I agree with the idea that we should not test this with the in-memory storage. How about using `STORAGE_MODE` and early return the test if the `storage_mode` is the in-memory storage? It would be appriciate that you provide some useful comments.

Reply:
Sorry. This is my mis-undestanding. `InMemoryStorage` has `read_trials_from_remote_storage()` and it can pass the test. So I fixed it to use just `STORAGE_MODES`.

I have modified `read_trials_from_remote_storage()` of `InMemoryStorage` to raise `NotImplemented` exception for investigation and wrongly used it for test...